core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java

```diff
@@ -198,39 +192,26 @@ private WriterTarget createWriterTarget(WriterTarget target)
             throw new IllegalArgumentException("Unhandled target type: " + target.getClass().getSimpleName());
         }
 
-        private PlanNode rewriteDeleteTableScan(PlanNode node, TableHandle handle)
+        private PlanNode rewriteUpdatingModifiedTableScan(PlanNode node, TableHandle tableHandle)
         {
             if (node instanceof TableScanNode) {
                 TableScanNode scan = (TableScanNode) node;
-                return new TableScanNode(
-                        scan.getId(),
-                        handle,
-                        scan.getOutputSymbols(),
-                        scan.getAssignments(),
-                        scan.getEnforcedConstraint(),
-                        scan.isForDelete());
-            }
-
-            if (node instanceof FilterNode) {
-                PlanNode source = rewriteDeleteTableScan(((FilterNode) node).getSource(), handle);
-                return replaceChildren(node, ImmutableList.of(source));
-            }
-            if (node instanceof ProjectNode) {
-                PlanNode source = rewriteDeleteTableScan(((ProjectNode) node).getSource(), handle);
-                return replaceChildren(node, ImmutableList.of(source));
-            }
-            if (node instanceof SemiJoinNode) {
-                PlanNode source = rewriteDeleteTableScan(((SemiJoinNode) node).getSource(), handle);
-                return replaceChildren(node, ImmutableList.of(source, ((SemiJoinNode) node).getFilteringSource()));
-            }
-            if (node instanceof JoinNode) {
-                JoinNode joinNode = (JoinNode) node;
-                if (joinNode.getType() == JoinNode.Type.INNER && isAtMostScalar(joinNode.getRight())) {
-                    PlanNode source = rewriteDeleteTableScan(joinNode.getLeft(), handle);
-                    return replaceChildren(node, ImmutableList.of(source, joinNode.getRight()));
+                if (scan.isForModify()) {
+                    return new TableScanNode(
+                            scan.getId(),
+                            tableHandle,
+                            scan.getOutputSymbols(),
+                            scan.getAssignments(),
+                            scan.getEnforcedConstraint(),
+                            scan.isForModify());
+                }
+                else {
+                    return scan;
                 }
             }
-            throw new IllegalArgumentException("Invalid descendant for DeleteNode: " + node.getClass().getName());
+            return node.replaceChildren(node.getSources().stream()
+                    .map(child -> rewriteUpdatingModifiedTableScan(child, tableHandle))
+                    .collect(toImmutableList()));
```

Comment:
In its current shape, the `rewriteDeleteTableScan` method only supports certain PlanNodes as descendants of `DeleteNode`. These are: scan, filter, project, semi join and join (but only of INNER type and with scalar subquery).

Contrarily, in the `rewriteUpdatingModifiedTableScan` method, all PlanNodes are supported and the rewrite is pushed down recursively.

Before we open that door, we should first make sure that the underlying plan is eligible for deleting. Currently, we have a lot of plan transformations, like join flipping (CBO) or AddExchanges that don't care if they are in the context of delete or not. (in fact, the only delete-aware transformation is semi-join to inner-join rewrite) As a consequence, it is possible that the delete plan is no more valid at this point.

I suggest that we do the validation at the point of this rewrite. It requires to:
1. push the rewrite only into the leftmost branch of the plan
2. fail in case we find exchange which breaks the delete plan. Also, identify and catch other nodes which break the delete plan
3. in case when the proper TableScanNode wasn't found or the plan is broken, throw exception other than `IllegalArgumentException`
4. and as a long-term goal, revisit PlanOptimizers and optimizer rules and make them delete-aware.

Reply:
Before this PR, the `TableScanNode` chosen as the target of DELETE was the first `TableScanNode` encountered by `rewriteDeleteTableScan`.  After this change, the `TableScanNode` chosen is the one referring to the `Table` object registered in `Analysis.setUpdateType()`, which is called by `StatementAnalyzer.visitDelete()`.  So it seems to me that after this change, we will always find the right `TableScanNode`, even if optimizations rearrange the plan tree in arbitrary ways.  It makes sense to change the exception raised if we don't find the `TableScanNode`

The concern you raise - - that EXCHANGE or other nodes might result in an invalid DELETE plan - - are independent of how we find the appropriate `TableScanNode`, right?  That is, all the changes in this PR do is pick the `TableScanNode` target in a more reliable way.  This PR doesn't change what optimizations will be applied to the plan.

Can you amplify on "identify and catch other nodes which break the delete plan"?  I'm not sure how it should be done.  Do you believe there are existing optimizations that will break the DELETE plan?  If so, I'd like to write tests for them. 
 Are there other validation routines that could be used as examples to learn from?

Reply:
@kasiafi and I chatted, and I now understand the concern she is raising.  She thinks the new way of finding the `TableScanNode` is an improvement.  Her concern is with `rewriteUpdatingModifiedTableScan`.  The code it replaces will not find the `TableScanNode` if optimizations insert an `ExchangeNode` between the root and the `TableScanNode`, and so the query will fail.  That's a good thing, because DELETE requires the `TableScanNode` to be on the same node as writes the delete records.

My replacement code will find the `TableScanNode` no matter where it is in the tree, and that means it may allow illegal DELETE plans.  @martint showed her a DELETE query which was illegal because it interposed an `ExchangeNode` between the `DeleteNode` and the `TableScanNode`.  I'll ask @martint about it.

As she indicates, to restore the previous functionality for DELETE, `rewriteUpdatingModifiedTableScan` should only traverse the left-hand children, and restrict plan nodes to scan, filter, project, semi join and inner join nodes.  SQL UPDATE has the same node processing restrictions as DELETE, so will use the same traversal to find the `TableScanNode`.

I believe SQL MERGE is different - - it has an explicit redistribute of the result of the RIGHT JOIN, so finding the `TableScanNode` anywhere in the tree seems appropriate for MERGE.